xdp-dump/xdpdump.c

```diff
@@ -99,17 +113,115 @@ static struct prog_option xdpdump_options[] = {
 };
 
 struct perf_handler_ctx {
-	uint64_t                   missed_events;
-	uint64_t                   captured_packets;
-	uint64_t                   epoch_delta;
-	struct dumpopt            *cfg;
-	pcap_t                    *pcap;
-	pcap_dumper_t             *pcap_dumper;
+	uint64_t               missed_events;
+	uint64_t               last_missed_events;
+	uint64_t               captured_packets;
+	uint64_t               epoch_delta;
+	uint64_t               packet_id;
+	uint64_t               cpu_packet_id[MAX_CPUS];
+	struct dumpopt        *cfg;
+	pcap_t                *pcap;
+	pcap_dumper_t         *pcap_dumper;
+	struct xpcapng_dumper *pcapng_dumper;
 };
 
 bool    exit_xdpdump;
 pcap_t *exit_pcap;
 
+/*****************************************************************************
+ * get_if_speed()
+ *****************************************************************************/
+static uint64_t get_if_speed(struct iface *iface)
+{
+#define MAX_MODE_MASKS 10
+
+	int                                  fd;
+	struct ifreq                         ifr;
+	struct {
+		struct ethtool_link_settings req;
+		uint32_t                     modes[3 * MAX_MODE_MASKS];
+	} ereq;
+
+	if (iface == NULL)
+		return 0;
+
+	/* Open socket, and initialize structures. */
+	fd = socket(AF_INET, SOCK_DGRAM, 0);
+	if (fd < 0)
+		return 0;
+
+	memset(&ereq, 0, sizeof(ereq));
+	ereq.req.cmd = ETHTOOL_GLINKSETTINGS;
+
+	memset(&ifr, 0, sizeof(ifr));
+	strncpy(ifr.ifr_name, iface->ifname, sizeof(ifr.ifr_name) - 1);
+	ifr.ifr_data = (void *)&ereq;
+
+	/* First query the kernel to see how many masks we need to ask for. */
+	if (ioctl(fd, SIOCETHTOOL, &ifr) != 0)
```

Comment:
Since we're dependent on a 5.7+ kernel anyway, why not use the ethtool netlink interface?

Reply:
We could, be do we want to pull in all the TLV parsing for Netlink? I thought this would be the easiest way. At the time I implemented this is still how ethtool does it.

Reply:
I guess we can change this if/when we add interface alias support :)